lib/fast_jsonapi/serialization_core.rb

```diff
@@ -82,11 +82,14 @@ def relationships_hash(record, relationships = nil, params = {})
         relationships = relationships_to_serialize if relationships.nil?
 
         relationships.each_with_object({}) do |(_k, relationship), hash|
-          name = relationship[:key]
-          empty_case = relationship[:relationship_type] == :has_many ? [] : nil
-          hash[name] = {
-            data: ids_hash_from_record_and_relationship(record, relationship, params) || empty_case
-          }
+          conditional_proc = relationship[:conditional_proc]
+          if conditional_proc.blank? || conditional_proc.call(record, params)
```

Comment:
I'm trying to abstract the relationship hash out into it's own class, much like we did in #256, but I'm having a hard time figuring out how to split some of the methods. For instance, `id_hash` would be used inside the `Relationship` class, but is a class method of `SerializationCore`, which is the class that calls the `Relationship`, so it seems a bit like it violates separation of concerns. There are other methods like that too.

@christophersansone since we were discussing the abstraction in the other PR that was merged in, any suggestions for something like that?

Reply:
@TrevorHinesley Thanks for your help here and for bringing me into the fold.  I'm looking at the relationship code for the first time in a while here, and it has definitely grown more complex.  The idea of extracting `Relationship` just like `Attribute` would be awesome here.

My approach would be to start at `relationships_hash` and make it look just like `attributes_hash`.  It looks like a lot of the methods called within are not called anywhere else, so it should go well.  The only method that appears to be reused elsewhere from my initial read is `id_hash`: it is used by both relationships and the record itself.  However, that method has a few different conditionals, and when the record calls it, it passes in that third parameter to change the behavior.  If we cleanly separated concerns here, I would just give `Relationship` its own `id_hash` method that did exactly what it needed, and just leave the existing `id_hash` alone for now for everyone else to use.

Eventually it looks like if we succeed here, we might eventually wind up with classes like these for `Record`, `Link`, `Include`, etc.  It would definitely make the code easier to reason about and should make most features easier to implement.